src/ppm/decoder.rs

```diff
@@ -0,0 +1,87 @@
+use std::io::Read;
+use std::io::BufReader;
+
+use color::{ColorType};
+use image::{DecodingResult, ImageDecoder, ImageResult, ImageError};
+
+/// PPM decoder
+pub struct PPMDecoder<R> {
+    reader: BufReader<R>,
+    width: u32,
+    height: u32,
+}
+
+impl<R: Read> PPMDecoder<R> {
+    /// Create a new decoder that decodes from the stream ```r```
+    pub fn new(r: R) -> ImageResult<PPMDecoder<R>> {
+        let mut buf = BufReader::new(r);
+        try!(PPMDecoder::read_next_string(&mut buf)); // Skip P6
+        let width = try!(PPMDecoder::read_next_u32(&mut buf));
+        let height = try!(PPMDecoder::read_next_u32(&mut buf));
+        try!(PPMDecoder::read_next_u32(&mut buf)); // Skip maxwhite
+        Ok(PPMDecoder {
+            reader: buf,
+            width: width,
+            height: height,
+        })
+    }
+
+  fn read_next_string(reader: &mut BufReader<R>) -> ImageResult<String> {
+      let mut bytes = Vec::new();
+      loop {
+          let mut buffer = vec![0 as u8];
```

Comment:
Wouldn't it be better to use a simple [u8] here? (e.g `let mut buffer = [0u8]`) Creating and dropping an 1-byte vec for each byte isn't going to be great for performance.